presto-main/src/main/java/io/prestosql/operator/scalar/ArraySubscriptOperator.java

```diff
@@ -77,7 +77,8 @@ else if (elementType.getJavaType() == Slice.class) {
             methodHandle = METHOD_HANDLE_SLICE;
         }
         else {
-            methodHandle = METHOD_HANDLE_OBJECT;
+            methodHandle = METHOD_HANDLE_OBJECT.asType(
+                    METHOD_HANDLE_OBJECT.type().changeReturnType(elementType.getJavaType()));
```

Comment:
I was planning to add suggestion to add extra casting in the generated code, as I was expecting array subscript is not only affected function.
For example I expected `MapSubscriptOperator` to have same issue. But for some reason I do not reproduce the problem with 
```
CREATE TABLE test (col MAP(VARCHAR,TIMESTAMP(7)));
```

interesting.

Reply:
I had the cast you suggested until I realized that `MapSubscriptOperator` doesn't have the bug. Then I copied the current solution from that class.

Another option I considered early on was to perform this same `MethodType` conversion in `generateFullInvocation`. I did it wrong at first, and now I think it's not quite as clean a solution. Yes, it would prevent this exact error in the future, but it seems like a better idea to trust each SQL function to correctly set its return type. (I've pushed this change as 3c8aac95.)

Maybe this should be in a doc comment on `SqlScalarFunction.specialize`.

Reply:
Should we maybe that method handle return type match the type in `FunctionRegistry.specializeScalarFunction`?

Reply:
That sounds reasonable to me. Let's consider this as a follow up maybe?

Reply:
+!